tests/pytests/unit/fileclient/test_fileclient_cache.py

```diff
@@ -0,0 +1,342 @@
+                    alt_cachedir, "files", saltenv, SUBDIR, subdir_file
+                )
+                # Double check that the content of the cached file
+                # identifies it as being from the correct saltenv. The
+                # setUp function creates the file with the name of the
+                # saltenv mentioned in the file, so a simple 'in' check is
+                # sufficient here. If opening the file raises an exception,
+                # this is a problem, so we are not catching the exception
+                # and letting it be raised so that the test fails.
+                with salt.utils.files.fopen(cache_loc) as fp_:
+                    content = fp_.read()
+                log.debug("cache_loc = %s", cache_loc)
+                log.debug("content = %s", content)
+                assert subdir_file in content
+                assert SUBDIR in content
+                assert saltenv in content
+
+
+def test_cache_dir_with_alternate_cachedir_and_relative_path(mocked_opts, minion_opts):
+    """
+    Ensure entire directory is cached to correct location when an alternate
+    cachedir is specified and that cachedir is a relative path
+    """
+    patched_opts = minion_opts.copy()
+    patched_opts.update(mocked_opts)
+    alt_cachedir = "foo"
+
+    with patch.dict(fileclient.__opts__, patched_opts):
+        client = fileclient.get_file_client(fileclient.__opts__, pillar=False)
+        for saltenv in _saltenvs():
+            assert client.cache_dir(
+                "salt://{}".format(SUBDIR), saltenv, cachedir=alt_cachedir
+            )
+            for subdir_file in _subdir_files():
+                cache_loc = os.path.join(
+                    fileclient.__opts__["cachedir"],
+                    alt_cachedir,
+                    "files",
+                    saltenv,
+                    SUBDIR,
+                    subdir_file,
+                )
+                # Double check that the content of the cached file
+                # identifies it as being from the correct saltenv. The
+                # setUp function creates the file with the name of the
+                # saltenv mentioned in the file, so a simple 'in' check is
+                # sufficient here. If opening the file raises an exception,
+                # this is a problem, so we are not catching the exception
+                # and letting it be raised so that the test fails.
+                with salt.utils.files.fopen(cache_loc) as fp_:
+                    content = fp_.read()
+                log.debug("cache_loc = %s", cache_loc)
+                log.debug("content = %s", content)
+                assert subdir_file in content
+                assert SUBDIR in content
+                assert saltenv in content
+
+
+def test_cache_file(mocked_opts, minion_opts):
+    """
+    Ensure file is cached to correct location
+    """
+    patched_opts = {x: y for x, y in minion_opts.items()}
```

Comment:
```suggestion
    patched_opts = minion_opts.copy()
```
Sorry didn't want to sound repetitive on my previous review.